README.md
[![Latest Stable Version](http://poser.pugx.org/nguyenanhung/template-backend-package/v)](https://packagist.org/packages/template-backend-package) [![Total Downloads](http://poser.pugx.org/nguyenanhung/template-backend-package/downloads)](https://packagist.org/packages/nguyenanhung/template-backend-package) [![Latest Unstable Version](http://poser.pugx.org/nguyenanhung/template-backend-package/v/unstable)](https://packagist.org/packages/nguyenanhung/template-backend-package) [![License](http://poser.pugx.org/nguyenanhung/template-backend-package/license)](https://packagist.org/packages/nguyenanhung/template-backend-package) [![PHP Version Require](http://poser.pugx.org/nguyenanhung/template-backend-package/require/php)](https://packagist.org/packages/nguyenanhung/template-backend-package)

# Template start Backend API Package

Template for repository Backend API - Basic, Simple and Lightweight

## Use this Template

First, you can `Use this template` for new project: [Use this template](https://github.com/nguyenanhung/template-backend-package/generate)

Second, clone your project to your to path in your machine

Finally, your edit file `composer.json` in root folder of project

```json
{
    "type": "library",
    "name": "nguyenanhung/template-backend-package",
    "description": "Structure Repository Template for Backend API Service Package",
    "keywords": [
        "backend",
        "helper",
        "library",
        "php"
    ],
    "homepage": "https://github.com/nguyenanhung/template-backend-package",
    "license": "MIT",
    "minimum-stability": "stable",
    "authors": [
        {
            "name": "Nguyen An Hung",
            "email": "dev@nguyenanhung.com",
            "homepage": "https://nguyenanhung.com",
            "role": "Developer"
        }
    ],
    "require": {
        "php": "^5.6 || ^7.0 || ^8.0",
        "ext-curl": "*",
        "ext-json": "*",
        "ext-mbstring": "*",
        "nguyenanhung/framework": "^1.0",
        "lcobucci/jwt": "^4.1 || ^4.0 || ^3.4.6"
    },
    "require-dev": {
        "nguyenanhung/seo": "^3.0 || ^2.0",
        "nguyenanhung/image": "^3.0 || ^2.0",
        "nguyenanhung/upload": "^2.0",
        "nguyenanhung/mailer-sdk": "^3.0",
        "tramtro/make-some-noise": "^1.0"
    },
    "autoload": {
        "psr-4": {
            "nguyenanhung\\Backend\\Your_Project\\": "src/"
        },
        "files": [
            "helpers/helpers.php"
        ]
    }
}
```

Replace name space `REPLACE_FOR_YOUR` to Project Backend space, example: `Google`. After change namespace, project namespace same `"nguyenanhung\\Backend\\Google\\": "src/"`

Finished, your can writing new awesome helper and library now time.

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
|-------------|----------------------|------------------|---------------|
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |

From Vietnam with Love <3

CHANGELOG.md
# CHANGELOG

## [1.0.5] - 2023/02/28

### Fixed

- [x] Upgrade vendor package, use `nguyenanhung/framework`

## [1.0.4] - 2022/08/07

### Fixed

- [x] Upgrade vendor package

## [1.0.3] - 2022/07/20

### Fixed

- [x] Add Validation

## [1.0.2] - 2022/07/04

### Fixed

- [x] Add Test

## [1.0.1] - 2022/07/02

### Fixed

- [x] Fixed Helper

## [1.0.0] - 2022/07/02

### Added

- [x] Create Project Structure

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
|-------------|----------------------|------------------|---------------|
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |


test/test.php
<?php
/**
 * Project template-backend-package
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 02/07/2022
 * Time: 00:19
 */

src/Base/BaseCore.php
<?php

namespace nguyenanhung\Backend\Your_Project\Base;

use nguyenanhung\MyDebug\Benchmark;
use nguyenanhung\MyDebug\Logger;
use nguyenanhung\MyCache\Cache;
use nguyenanhung\MyRequests\MyRequests;
use nguyenanhung\Backend\Your_Project\Helper\Helper;

/**
 * Class BaseCore
 *
 * @package   nguyenanhung\Backend\Your_Project\Base
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class BaseCore
{
    const VERSION = '1.0.5';
    const KEY_DATABASE = 'DATABASE';
    const KEY_OPTIONS = 'OPTIONS';
    const KEY_AUTH = 'AUTH';
    const KEY_CONFIG = 'CONFIG';

    /** @var \nguyenanhung\Backend\Your_Project\Helper\Helper */
    protected $helper;
    /** @var \nguyenanhung\MyDebug\Benchmark */
    protected $benchmark;
    /** @var \nguyenanhung\MyDebug\Logger */
    protected $logger;
    /** @var \nguyenanhung\MyCache\Cache */
    protected $cache;
    /** @var \nguyenanhung\MyRequests\MyRequests */
    protected $requests;
    /** @var string $cachePrefix */
    protected $cachePrefix;
    /** @var array SDK Config Options */
    protected $options;
    /** @var array SDK Config */
    protected $sdkConfig;
    /** @var bool Set Response is Object */
    protected $responseIsObject;
    /** @var bool Set Response is JSON */
    protected $responseIsJson;
    /** @var mixed Response Data */
    protected $response;
    /** @var mixed $inputData */
    protected $inputData;

    /**
     * BaseCore constructor.
     *
     * @param array $options
     *
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     */
    public function __construct(array $options = array())
    {
        $this->helper = new Helper();
        $this->logger = new Logger();
        $this->benchmark = new Benchmark();
        $this->requests = new MyRequests();
        $this->cache = new Cache();
        if (isset($options['debugStatus']) && $options['debugStatus'] === true) {
            $this->logger->setDebugStatus(true);
            $this->cache->setDebugStatus(true);
            if (isset($options['debugLevel']) && !empty($options['debugLevel'])) {
                $this->logger->setGlobalLoggerLevel($options['debugLevel']);
                $this->cache->setDebugLevel($options['debugLevel']);
            }
            if (isset($options['loggerPath']) && !empty($options['loggerPath'])) {
                $this->logger->setLoggerPath($options['loggerPath']);
                $this->cache->setDebugLoggerPath($options['loggerPath']);
            }
            $this->logger->setLoggerSubPath(__CLASS__);
            $this->logger->setLoggerFilename('Log-' . date('Y-m-d') . '.log');
        }
        if (isset($options['cachePath'])) {
            $this->cache->setCachePath($options['cachePath']);
        }
        if (isset($options['cacheTtl'])) {
            $this->cache->setCacheTtl(300);
        }
        if (isset($options['cacheDriver'])) {
            $this->cache->setCacheDriver($options['cacheDriver']);
        }
        if (isset($options['cacheFileDefaultChmod'])) {
            $this->cache->setCacheDefaultChmod($options['cacheFileDefaultChmod']);
        }
        if (isset($options['cacheSecurityKey'])) {
            $this->cache->setCacheSecurityKey($options['cacheSecurityKey']);
        }
        $this->cache->__construct();
        $this->options = $options;
    }

    /**
     * Function getVersion
     *
     * @return string
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 40:11
     */
    public function getVersion()
    {
        return self::VERSION;
    }

    /**
     * Function setSdkConfig
     *
     * @param array $sdkConfig
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 37:01
     */
    public function setSdkConfig(array $sdkConfig)
    {
        $this->sdkConfig = $sdkConfig;
        if (isset($this->db)) {
            $this->db->setSdkConfig($this->sdkConfig);
            if (isset($this->sdkConfig[self::KEY_DATABASE])) {
                $this->db->setDatabase($this->sdkConfig[self::KEY_DATABASE]);
            }
            $this->db->__construct($this->sdkConfig[self::KEY_OPTIONS]);
        }

        return $this;
    }

    /**
     * Function setInputData
     *
     * @param $inputData
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 37:37
     */
    public function setInputData($inputData)
    {
        $this->inputData = $inputData;
        $this->logger->debug('InputData', 'Received Input Data: ', $this->inputData);

        return $this;
    }

    /**
     * Function setResponseIsObject
     *
     * @param bool $responseIsObject
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 48:40
     */
    public function setResponseIsObject($responseIsObject)
    {
        $this->responseIsObject = $responseIsObject;

        return $this;
    }

    /**
     * Function setResponseIsJson
     *
     * @param bool $responseIsJson
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 49:31
     */
    public function setResponseIsJson($responseIsJson)
    {
        $this->responseIsJson = $responseIsJson;

        return $this;
    }

    /**
     * Function getResponse
     *
     * @return mixed
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 37:56
     */
    public function getResponse()
    {
        if ($this->responseIsObject === true) {
            $this->response = $this->helper->arrayToObject($this->response);
        }
        if ($this->responseIsJson === true) {
            $this->response = json_encode($this->response);
        }

        return $this->response;
    }
}


src/Http/BaseHttp.php
<?php

namespace nguyenanhung\Backend\Your_Project\Http;

use nguyenanhung\Backend\Your_Project\Base\BaseCore;
use nguyenanhung\Backend\Your_Project\Database\Database;

/**
 * Class BaseHttp
 *
 * @package   nguyenanhung\Backend\Your_Project\Http
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class BaseHttp extends BaseCore
{
    const EXIT_CODE = array(
        'success'           => 0,
        'contentIsEmpty'    => 1,
        'invalidParams'     => 2,
        'invalidSignature'  => 3,
        'outdatedSignature' => 4,
        'invalidService'    => 5,
        'paramsIsEmpty'     => 6
    );

    /** @var \nguyenanhung\Backend\Your_Project\Database\Database */
    protected $db;

    /**
     * BaseHttp constructor.
     *
     * @param array $options
     *
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     */
    public function __construct(array $options = array())
    {
        parent::__construct($options);
        $this->logger->setLoggerSubPath(__CLASS__);
        $this->db = new Database($options);
    }

    protected function formatInputStartDate($inputData = array())
    {
        if (isset($inputData['begin_date'])) {
            $startDate = $inputData['begin_date'];
        } elseif (isset($inputData['begindate'])) {
            $startDate = $inputData['begindate'];
        } elseif (isset($inputData['start_date'])) {
            $startDate = $inputData['start_date'];
        } else {
            $startDate = null;
        }

        return $startDate;
    }

    protected function formatInputEndDate($inputData = array())
    {
        if (isset($inputData['end_date'])) {
            $endDate = $inputData['end_date'];
        } elseif (isset($inputData['enddate'])) {
            $endDate = $inputData['enddate'];
        } else {
            $endDate = null;
        }

        return $endDate;
    }

    protected function formatInputUsername($inputData = array())
    {
        if (isset($inputData['username'])) {
            $res = $inputData['username'];
        } elseif (isset($inputData['nickname'])) {
            $res = $inputData['nickname'];
        } elseif (isset($inputData['account'])) {
            $res = $inputData['account'];
        } elseif (isset($inputData['acc'])) {
            $res = $inputData['acc'];
        } else {
            $res = null;
        }

        return $res;
    }

    protected function formatInputSignature($inputData = array())
    {
        if (isset($inputData['signature'])) {
            $res = $inputData['signature'];
        } elseif (isset($inputData['signal'])) {
            $res = $inputData['signal'];
        } elseif (isset($inputData['token'])) {
            $res = $inputData['token'];
        } elseif (isset($inputData['secret_token'])) {
            $res = $inputData['secret_token'];
        } else {
            $res = null;
        }

        return $res;
    }
}


src/Config/DataRepository.php
<?php

namespace nguyenanhung\Backend\Your_Project\Config;

/**
 * Class DataRepository
 *
 * @package   nguyenanhung\Backend\Your_Project\Config
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class DataRepository
{
    const CONFIG_PATH = 'config';
    const CONFIG_EXT = '.php';

    /**
     * Hàm lấy nội dung config được quy định trong thư mục config
     *
     * @author: 713uk13m <dev@nguyenanhung.com>
     * @time  : 9/28/18 14:47
     *
     * @param string $configName Tên file config
     *
     * @return array|mixed
     */
    public static function getData($configName)
    {
        $path = __DIR__ . DIRECTORY_SEPARATOR . self::CONFIG_PATH . DIRECTORY_SEPARATOR . $configName . self::CONFIG_EXT;
        if (is_file($path) && file_exists($path)) {
            return require $path;
        }

        return array();
    }
}


src/Http/WebServiceAccount.php
<?php

namespace nguyenanhung\Backend\Your_Project\Http;

/**
 * Class WebServiceAccount
 *
 * @package   nguyenanhung\Backend\Your_Project\Http
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class WebServiceAccount extends BaseHttp
{
    /**
     * WebServiceAccount constructor.
     *
     * @param array $options
     *
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     */
    public function __construct(array $options = array())
    {
        parent::__construct($options);
        $this->logger->setLoggerSubPath(__CLASS__);
    }

    public function register()
    {
        $data = array(
            'code'        => self::EXIT_CODE['success'],
            'description' => 'Success'
        );
        $this->response = $data;

        return $this;
    }

    public function login()
    {
        $data = array(
            'code'        => self::EXIT_CODE['success'],
            'description' => 'Success'
        );
        $this->response = $data;

        return $this;
    }

    public function logout()
    {
        $data = array(
            'code'        => self::EXIT_CODE['success'],
            'description' => 'Success'
        );
        $this->response = $data;

        return $this;
    }

    public function updatePassword()
    {
        $data = array(
            'code'        => self::EXIT_CODE['success'],
            'description' => 'Success'
        );
        $this->response = $data;

        return $this;
    }

    public function forgotPassword()
    {
        $data = array(
            'code'        => self::EXIT_CODE['success'],
            'description' => 'Success'
        );
        $this->response = $data;

        return $this;
    }
}


src/Database/Traits/SignatureTable.php
<?php

namespace nguyenanhung\Backend\Your_Project\Database\Traits;

/**
 * Trait SignatureTable
 *
 * @package   nguyenanhung\Backend\Your_Project\Database\Traits
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
trait SignatureTable
{
    /**
     * Function getUserSignature
     *
     * @param $nickname
     * @param $role
     *
     * @return array|bool|\Illuminate\Support\Collection|mixed|object|string|null
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 28/02/2023 35:51
     */
    public function getUserSignature($nickname = '', $role = null)
    {
        $table = 'data_signature';
        $cacheKey = $this->cachePrefix . __FUNCTION__ . hash('md5', $table . $nickname . $role);
        if ($this->cache->has($cacheKey)) {
            $result = $this->cache->get($cacheKey);
        } else {
            $DB = $this->connection();
            $wheres = array();
            $wheres['nickname'] = array('field' => 'nickname', 'operator' => $DB::OPERATOR_EQUAL_TO, 'value' => $nickname);
            $wheres['status'] = array('field' => 'status', 'operator' => $DB::OPERATOR_EQUAL_TO, 'value' => 1);
            if ($role !== null) {
                $wheres['status'] = array('field' => 'role', 'operator' => $DB::OPERATOR_EQUAL_TO, 'value' => $role);
            }
            $select = array('nickname', 'signature', 'role');
            $result = $DB->setTable($table)->getInfo($wheres, 'id', null, $select);
            $this->cache->save($cacheKey, $result);
            $DB->disconnect();
            unset($DB);
        }

        return $result;
    }
}


src/Database/Database.php
<?php

namespace nguyenanhung\Backend\Your_Project\Database;

use nguyenanhung\Backend\Your_Project\Base\BaseCore;
use nguyenanhung\Backend\Your_Project\Database\Traits\SignatureTable;
use nguyenanhung\MyDatabase\Model\BaseModel;

/**
 * Class Database
 *
 * @package   nguyenanhung\Backend\Your_Project\Database
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class Database extends BaseCore
{
    use SignatureTable;

    /** @var array $database */
    protected $database;

    /**
     * Database constructor.
     *
     * @param array $options
     *
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     */
    public function __construct(array $options = array())
    {
        parent::__construct($options);
        $this->logger->setLoggerSubPath(__CLASS__);
    }

    /**
     * Function setDatabase
     *
     * @param $database
     *
     * @return $this
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 38:16
     */
    public function setDatabase($database)
    {
        $this->database = $database;

        return $this;
    }

    /**
     * Function connection
     *
     * @return \nguyenanhung\MyDatabase\Model\BaseModel
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 22/06/2022 40:58
     */
    public function connection()
    {
        $DB = new BaseModel();
        $DB->debugStatus = $this->options['debugStatus'];
        $DB->debugLevel = $this->options['debugLevel'];
        $DB->debugLoggerPath = $this->options['loggerPath'];
        $DB->debugLoggerFilename = 'Log-' . date('Y-m-d') . '.log';
        $DB->setDatabase($this->database);
        $DB->__construct($this->database);

        return $DB;
    }

    /**
     * Function checkExitsRecord
     *
     * @param $wheres
     * @param $tableName
     *
     * @return bool
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 07/08/2022 13:25
     */
    public function checkExitsRecord($wheres, $tableName)
    {
        $DB = $this->connection();
        $DB->setTable($tableName);
        $result = $DB->checkExists($wheres);
        $DB->disconnect();
        unset($DB);

        return $result === 1;
    }
}
